File: server/src/main/java/de/zalando/zally/rule/SupportedRulesController.java
package de.zalando.zally.rule;

import de.zalando.zally.dto.RuleDTO;
import de.zalando.zally.dto.RulesListDTO;
import de.zalando.zally.dto.SeverityBinder;
import de.zalando.zally.rule.api.Rule;
import de.zalando.zally.rule.api.Severity;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.WebDataBinder;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.InitBinder;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

@CrossOrigin
@RestController
public class SupportedRulesController {

    private final RulesManager rules;
    private final RulesPolicy rulesPolicy;

    @Autowired
    public SupportedRulesController(RulesManager rules, RulesPolicy rulesPolicy) {
        this.rules = rules;
        this.rulesPolicy = rulesPolicy;
    }

    @InitBinder
    public void initBinder(final WebDataBinder binder) {
        binder.registerCustomEditor(Severity.class, new SeverityBinder());
    }

    @ResponseBody
    @GetMapping("/supported-rules")
    public RulesListDTO listSupportedRules(
        @RequestParam(value = "type", required = false) Severity typeFilter,
        @RequestParam(value = "is_active", required = false) Boolean isActiveFilter) {

        List<RuleDTO> filteredRules = rules
            .getRules()
            .stream()
            .filter(details -> filterByIsActive(details.getInstance(), isActiveFilter))
            .filter(details -> filterByType(details.getInstance(), typeFilter))
            .map(this::toDto)
            .sorted(comparing(RuleDTO::getType)
                .thenComparing(RuleDTO::getCode)
                .thenComparing(RuleDTO::getTitle))
            .collect(toList());

        return new RulesListDTO(filteredRules);
    }

    private boolean filterByIsActive(Rule rule, Boolean isActiveFilter) {
        boolean isActive = rulesPolicy.accepts(rule);
        return isActiveFilter == null || isActive == isActiveFilter;
    }

    private boolean filterByType(Rule rule, Severity typeFilter) {
        return typeFilter == null || typeFilter.equals(rule.getSeverity());
    }

    private RuleDTO toDto(RuleDetails details) {
        return new RuleDTO(
                details.getInstance().getTitle(),
                details.getInstance().getSeverity(),
                details.getRuleSet().url(details.getInstance()).toString(),
                details.getInstance().getId(),
                rulesPolicy.accepts(details.getInstance())
        );
    }
}

File: server/zally-server/src/main/kotlin/org/zalando/zally/rule/SupportedRulesController.kt
package org.zalando.zally.rule

import org.zalando.zally.core.RuleDetails
import org.zalando.zally.core.RulesManager
import org.zalando.zally.core.RulesPolicy
import org.zalando.zally.dto.RuleDTO
import org.zalando.zally.dto.RulesListDTO
import org.zalando.zally.dto.SeverityBinder
import org.zalando.zally.rule.api.Severity
import org.springframework.web.bind.WebDataBinder
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.InitBinder
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.RestController

/**
 * REST API for listing the rules supported by this server.
 */
@CrossOrigin
@RestController
class SupportedRulesController(private val rules: RulesManager, private val rulesPolicy: RulesPolicy) {

    /**
     * Registers serialization binding for Severity enum
     * @param binder the binder to configure
     */
    @InitBinder
    fun initBinder(binder: WebDataBinder) {
        binder.registerCustomEditor(Severity::class.java, SeverityBinder())
    }

    /**
     * GET /supported-rules implementation.
     * @param typeFilter filter rules to just this severity
     * @param isActiveFilter filter rules to just active ones
     * @return rules list ready for serialization
     */
    @ResponseBody
    @GetMapping("/supported-rules")
    fun listSupportedRules(
        @RequestParam(value = "type", required = false) typeFilter: Severity?,
        @RequestParam(value = "is_active", required = false) isActiveFilter: Boolean?
    ): RulesListDTO {

        val filteredRules = rules
            .rules
            .filter { filterByIsActive(it, isActiveFilter) }
            .filter { filterByType(it, typeFilter) }
            .map { this.toDto(it) }
            .sortedWith(
                compareBy(
                    RuleDTO::type,
                    RuleDTO::code,
                    RuleDTO::title
                )
            )

        return RulesListDTO(filteredRules)
    }

    private fun filterByIsActive(details: RuleDetails, isActiveFilter: Boolean?): Boolean {
        val isActive = rulesPolicy.accepts(details.rule)
        return isActiveFilter == null || isActive == isActiveFilter
    }

    private fun filterByType(details: RuleDetails, typeFilter: Severity?): Boolean {
        return typeFilter == null || typeFilter == details.rule.severity
    }

    private fun toDto(details: RuleDetails): RuleDTO {
        return RuleDTO(
            details.rule.title,
            details.rule.severity,
            details.ruleSet.url(details.rule).toString(),
            details.rule.id,
            rulesPolicy.accepts(details.rule)
        )
    }
}
